xla/client/lib/math.cc

```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
+      // that is based on splitting the first quarter of the complex
+      // plane into 11 regions and providing its own approximations
+      // to each region. The evaluation of arcsin in other quarters
+      // is resolved by using identities
+      //   arcsin(-z) == -arcsin(z)
+      //   arcsin(conj(z)) == conj(arcsin(z))
+      //
+      // When considering the evaluation of arcsin real and imaginary
+      // parts separately, the 11 regions by Hull et al can be reduced
+      // to 3 regions for the real part and 6 regions for the
+      // imaginary part. In the following, the corresponding
+      // modification of the Hull et al algorithm is provided. It is
+      // validated against the mpmath.mp.asin that implements the
+      // original Hull et al algorithm.
+
+      auto dtype = primitive_util::ComplexComponentType(shape.element_type());
+
+      double safe_min_, safe_max_;
+      switch (dtype) {
+        case F64:
+          safe_min_ = std::sqrt(std::numeric_limits<double>::min()) * 4;
+          safe_max_ = std::sqrt(std::numeric_limits<double>::max()) / 8;
+          break;
+        case F32:
+          safe_min_ = std::sqrt(std::numeric_limits<float>::min()) * 4;
+          safe_max_ = std::sqrt(std::numeric_limits<float>::max()) / 8;
+          break;
+        default:
+          safe_min_ = safe_max_ = 0.0;
+          assert(false);  // unreachable
+      }
+      auto signed_x = Real(z);
+      auto signed_y = Imag(z);
+      auto x = Abs(signed_x);
+      auto y = Abs(signed_y);
+
+      auto safe_min = ScalarLike(x, safe_min_);
+      auto safe_max = ScalarLike(x, safe_max_);
+      auto safe_max_m6 = ScalarLike(x, safe_max_ * 1e-6);
+      auto safe_max_p12 = ScalarLike(x, safe_max_ * 1e12);
+      auto safe_max_p2 = ScalarLike(x, safe_max_ * 1e2);
+      auto safe_max_opt = Select(Lt(x, safe_max_p12), safe_max_m6, safe_max_p2);
+      auto y_ge_max_opt = Ge(y, safe_max_opt);
+
+      auto zero = ScalarLike(x, 0);
+      auto one = ScalarLike(x, 1);
+      auto none = ScalarLike(x, -1);
+      auto half = ScalarLike(x, 0.5);
+      auto log2 = ScalarLike(x, std::log(2.0));
+      auto xp1 = Add(x, one);
+      auto xm1 = Add(x, none);
+      auto yy = Mul(y, y);
+      auto half_yy = Mul(half, yy);
+      auto r = Hypot(xp1, y);
+      auto s = Hypot(xm1, y);
+      auto a = Mul(half, Add(r, s));
+      auto ap1 = Add(a, one);
+      auto half_apx = Mul(half, Add(a, x));
+      auto rpxp1 = Add(r, xp1);
+      auto spxm1 = Add(s, xm1);
+      auto smxm1 = Sub(s, xm1);
+      auto y_is_finite = Not(IsPosInf(y));
+      auto max_xy = Select(Gt(x, y), x, y);
+
+      auto y1 = Select(
+          Gt(max_xy, safe_max), y,
+          Select(
+              Le(x, one), Sqrt(Mul(half_apx, Add(Div(yy, rpxp1), smxm1))),
+              Mul(y, Sqrt(Add(Div(half_apx, rpxp1), Div(half_apx, spxm1))))));
+      auto real = Atan2(signed_x, y1);
+
+      auto am1 = Select(
```

Comment:
Is this `a-1` with greater precision? If so a comment would be useful.

Reply:
`a - 1` approximation is now described in math_imph.h